keyboards/moon/matrix.c

```diff
@@ -0,0 +1,202 @@
+ *       `----------'                   `----------'
+ */
+
+// PCA9555 slave addresses
+#define IC1 0x20
+#define IC2 0x21
+
+// PCA9555 column pin masks
+#define PORT0_COLS_MASK 0b11111000
+#define PORT1_COLS_MASK 0b00111111
+#define COLS_MASK 0b0000011111111111
+
+#if (MATRIX_COLS <= 8)
+#  define print_matrix_header() print("\nr/c 01234567\n")
+#  define print_matrix_row(row) print_bin_reverse8(matrix_get_row(row))
+#  define matrix_bitpop(i) bitpop(matrix[i])
+#  define ROW_SHIFTER ((uint8_t)1)
+#elif (MATRIX_COLS <= 16)
+#  define print_matrix_header() print("\nr/c 0123456789ABCDEF\n")
+#  define print_matrix_row(row) print_bin_reverse16(matrix_get_row(row))
+#  define matrix_bitpop(i) bitpop16(matrix[i])
+#  define ROW_SHIFTER ((uint16_t)1)
+#elif (MATRIX_COLS <= 32)
+#  define print_matrix_header() print("\nr/c 0123456789ABCDEF0123456789ABCDEF\n")
+#  define print_matrix_row(row) print_bin_reverse32(matrix_get_row(row))
+#  define matrix_bitpop(i) bitpop32(matrix[i])
+#  define ROW_SHIFTER ((uint32_t)1)
+#endif
+
+/* matrix state(1:on, 0:off) */
+static matrix_row_t raw_matrix[MATRIX_ROWS];  // raw values
+static matrix_row_t matrix[MATRIX_ROWS];      // debounced values
+
+__attribute__((weak)) void matrix_init_quantum(void) { matrix_init_kb(); }
+
+__attribute__((weak)) void matrix_scan_quantum(void) { matrix_scan_kb(); }
+
+__attribute__((weak)) void matrix_init_kb(void) { matrix_init_user(); }
+
+__attribute__((weak)) void matrix_scan_kb(void) { matrix_scan_user(); }
+
+__attribute__((weak)) void matrix_init_user(void) {}
+
+__attribute__((weak)) void matrix_scan_user(void) {}
+
+inline uint8_t matrix_rows(void) { return MATRIX_ROWS; }
+
+inline uint8_t matrix_cols(void) { return MATRIX_COLS; }
+
+inline bool matrix_is_on(uint8_t row, uint8_t col) { return (matrix[row] & ((matrix_row_t)1 << col)); }
+
+inline matrix_row_t matrix_get_row(uint8_t row) { return matrix[row]; }
+
+void matrix_print(void) {
+  print_matrix_header();
+
+  for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
+    phex(row);
+    print(": ");
+    print_matrix_row(row);
+    print("\n");
+  }
+}
+
+uint8_t matrix_key_count(void) {
+  uint8_t count = 0;
+  for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+    count += matrix_bitpop(i);
+  }
+  return count;
+}
+
+static void init_i2c(void) {
+  pca9555_init(IC1);
+  pca9555_init(IC2);
+}
+
+static void init_pins(void) {
+  // init cols - IC2 port0 & IC2 port1 input
+  pca9555_set_config(IC2, PCA9555_PORT0, ALL_INPUT);
+  pca9555_set_config(IC2, PCA9555_PORT1, ALL_INPUT);
+
+  // init rows - IC1 port0 output
+  pca9555_set_config(IC1, PCA9555_PORT0, ALL_OUTPUT);
+  pca9555_set_output(IC1, PCA9555_PORT0, ALL_HIGH);
+}
+
+static void select_row(uint8_t row) {
+  // All rows are on the same IC and port
+  uint8_t mask = 1 << row;
+
+  // set active row low  : 0
+  // set other rows hi-Z : 1
+  pca9555_set_output(IC1, PCA9555_PORT0, ALL_HIGH & (~mask));
+}
+
+static uint32_t read_cols(void) {
```

Comment:
Columns is < 16 so this can safely use a smaller int.
```suggestion
static uint16_t read_cols(void) {
```

Reply:
Fixed.

I also added a small comment with the col/row pin mappings. I guess it could be handy at some point.